In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv('cancer.csv')

In [3]:
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   symmetry_mean  fractal_dimension_mean  radius_se  texture_se  perimeter_se  \
0         0.2419                 0.07871     1.0950      0.9053         8.589   
1         0.1812                 0.05667     0.5435      0.7339         3.398   
2         0.2069                 0.05999     0.7456      0.7869         4.585   
3         0.2597                 0.09744     0.4956      1.1560         3.445   
4         0.1809                 0.05883     0.7572      0.7813         5.438   

   area_se  smoothness_se  compactness_se  concavity_se  concave points_se  \
0   153.40       0.006399         0.04904       0.05373            0.01587   
1    74.08       0.005225         0.01308       0.01860            0.01340   
2    94.03       0.006150         0.04006       0.03832            0.02058   
3    27.23       0.009110         0.07458       0.05661            0.01867   
4    94.44       0.011490         0.02461       0.05688            0.01885   

   symmetry_se  fractal_dimension_se  radius_worst  texture_worst  \
0      0.03003              0.006193         25.38          17.33   
1      0.01389              0.003532         24.99          23.41   
2      0.02250              0.004571         23.57          25.53   
3      0.05963              0.009208         14.91          26.50   
4      0.01756              0.005115         22.54          16.67   

   perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
0           184.60      2019.0            0.1622             0.6656   
1           158.80      1956.0            0.1238             0.1866   
2           152.50      1709.0            0.1444             0.4245   
3            98.87       567.7            0.2098             0.8663   
4           152.20      1575.0            0.1374             0.2050   

   concavity_worst  concave points_worst  symmetry_worst  \
0           0.7119                0.2654          0.4601   
1           0.2416                0.1860          0.2750   
2           0.4504                0.2430          0.3613   
3           0.6869                0.2575          0.6638   
4           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN

In [4]:
df.shape

(569, 33)

In [5]:
df.drop(columns=['id','Unnamed: 32'],inplace=True)

In [6]:
df.diagnosis.unique()

array(['M', 'B'], dtype=object)

In [7]:
df.diagnosis.value_counts()

B    357
M    212
Name: diagnosis, dtype: int64

Since we have a small dataset with high features, we'll make use of **Random Forests Algorithm** to classify this data.

In [8]:
X = df.drop(columns=['diagnosis'])

In [9]:
y = df['diagnosis'] 

In [10]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.15, random_state = 45)

Data Set Characteristics:
    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

In [11]:
x_train.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
372       21.370         15.10          141.30     1386.0          0.10010   
341        9.606         16.84           61.64      280.5          0.08481   
108       22.270         19.67          152.80     1509.0          0.13260   
197       18.080         21.84          117.40     1024.0          0.07371   
258       15.660         23.20          110.20      773.5          0.11090   

     compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
372           0.15150         0.19320              0.12550         0.1973   
341           0.09228         0.08422              0.02292         0.2036   
108           0.27680         0.42640              0.18230         0.2556   
197           0.08642         0.11030              0.05778         0.1770   
258           0.31140         0.31760              0.13770         0.2495   

     fractal_dimension_mean  radius_se  texture_se  perimeter_se  area_se  \
372                 0.06183     0.3414      1.3090         2.407    39.06   
341                 0.07125     0.1844      0.9429         1.429    12.07   
108                 0.07039     1.2150      1.5450        10.050   170.00   
197                 0.05340     0.6362      1.3050         4.312    76.36   
258                 0.08104     1.2920      2.4540        10.120   138.50   

     smoothness_se  compactness_se  concavity_se  concave points_se  \
372       0.004426         0.02675       0.03437            0.01343   
341       0.005954         0.03471       0.05028            0.00851   
108       0.006515         0.08668       0.10400            0.02480   
197       0.005530         0.05296       0.06110            0.01444   
258       0.012360         0.05995       0.08232            0.03024   

     symmetry_se  fractal_dimension_se  radius_worst  texture_worst  \
372      0.01675              0.004367         22.69          21.84   
341      0.01750              0.004031         10.75          23.07   
108      0.03112              0.005037         28.40          28.01   
197      0.02140              0.005036         19.76          24.70   
258      0.02337              0.006042         19.85          31.64   

     perimeter_worst  area_worst  smoothness_worst  compactness_worst  \
372           152.10      1535.0           0.11920             0.2840   
341            71.25       353.6           0.12330             0.3416   
108           206.80      2360.0           0.17010             0.6997   
197           129.10      1228.0           0.08822             0.1963   
258           143.70      1226.0           0.15040             0.5172   

     concavity_worst  concave points_worst  symmetry_worst  \
372           0.4024               0.19660          0.2730   
341           0.4341               0.08120          0.2982   
108           0.9608               0.29100          0.4055   
197           0.2535               0.09181          0.2369   
258           0.6181               0.24620          0.3277   

     fractal_dimension_worst  
372                  0.08666  
341                  0.09825  
108                  0.09789  
197                  0.06558  
258                  0.10190

In [12]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators=200, bootstrap = True, max_features = 'sqrt')
model.fit(x_train, y_train)
y_pred = model.predict(x_test)
print('Accuracy of Random Forest on test set: {:.2f}'.format(model.score(x_test, y_test)))

Accuracy of Random Forest on test set: 0.94


In [13]:
data = [[14.12,19.3,91.7,655,0.1,0.1,0.08,0.05,0.1811,0.0627,0.4,1.2,2.8,40.33,0.0070,0.025,0.03,0.0011,0.020,0.0037,16,25,107,880,0.132,0.254,0.272,0.114,0.290,0.083]]

In [14]:
pred = model.predict(data)

In [15]:
pred[0]

'B'

In [16]:
import pickle

In [17]:
#model_cancer = pickle.dump(model,open('model_cancer.pkl','wb'))

In [18]:
x_train.columns

Index(['radius_mean', 'texture_mean', 'perimeter_mean', 'area_mean',
       'smoothness_mean', 'compactness_mean', 'concavity_mean',
       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',
       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',
       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',
       'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'perimeter_worst', 'area_worst', 'smoothness_worst',
       'compactness_worst', 'concavity_worst', 'concave points_worst',
       'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')